[cli] Fetch environment variables from project if `.env` is not present

MaxLeiter opened the pull request:
Previously, users would have to run `vc env pull` to fetch cloud environment variables into `.env`. After this PR, if no `.env` or `.build.env` file is present, environment variables will be pulled by `vc dev` from your Vercel Environment Variables settings, no file necessary.

leo commented:
Think about it like this:

1. Load Development Environment Variables from the cloud, into memory.
2. If it exists, read `.env` into memory, make it overwrite the cloud Environment Variables and only apply it to runtime.
3. If it exists, read `.env.build` into memory, make it overwrite the cloud Environment Variables and only apply it to build time.

MaxLeiter commented on packages/now-cli/src/commands/dev/dev.ts:
```diff
@@ -86,5 +87,5 @@ export default async function dev(
   process.once('SIGTERM', () => devServer.stop());
 
-  await devServer.start(...listen);
+  await devServer.start(client, link.project, ...listen);
```
@TooTallNate The problem with this approach is now `DevServer` needs an idea of Client and the project

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -2216,3 +2243,45 @@ function hasNewRoutingProperties(nowConfig: NowConfig) {
   );
 }
+
+async function fetchAndDecryptEnvVars(
```
Can we reuse the code from `vc env pull` instead of implementing again here?

This seems to be copy/pasted.

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -762,10 +771,28 @@ export default class DevServer {
+    const localBuildEnvVars = await this.getLocalEnv(
+      '.env.build',
+      nowConfigBuild.env
+    );
```
Any reason why these were changed from parallel to sequential reads?

MaxLeiter commented:
Addressed your comments @styfle

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -762,10 +769,28 @@ export default class DevServer {
+      : cloudEnvVariables;
+
+    // if local .env/.env.build file exists, don't use cloud variables
     const allEnv = { ...buildEnv, ...runEnv };
```
I think you need to change this logic because the build/run env vars only makes sense if there are files. If you are using the cloud, then it should be all or nothing.

Try something like this instead:

```js
let allEnv = { ...buildEnv, ...runEnv };
if (Object.keys(allEnv).length === 0) {
  allEnv = await getDecryptedEnvRecords(/*...*/);
}
```

styfle reviewed (commented):
We need to add an integration test for this new behavior in [./packages/now-cli/test/integration.js](https://github.com/vercel/vercel/blob/c80bb37e8d66eac83da40746cf1a15ba42d216d8/packages/now-cli/test/integration.js#L374).

TooTallNate commented:
@styfle It should be in `test/dev/integration.js` I think.

styfle commented:
@TooTallNate The dev tests are usually meant to create a deployment and test the paths between dev and production. But this test will be specific to the CLI behavior I think. That's why I linked to the previous `vercel dev` test we already have. I suppose this test could be implemented in either file.

TooTallNate commented:
Fair enough. I don't really have a preference then.

TooTallNate commented on packages/now-cli/test/integration.js:
```diff
@@ -244,7 +249,7 @@ test.after.always(async () => {
 
   // Make sure the token gets revoked
-  await execa(binaryPath, ['logout', ...defaultArgs]);
+  // await execa(binaryPath, ['logout', ...defaultArgs]);
```
Don't commit this part, otherwise tokens will be leftover on the test users 😬

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -602,6 +631,11 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
+  // await nowEnvRemove();
+  // await nowEnvRemoveWithArgs();
+  // await nowEnvRemoveWithNameOnly();
+  // await nowEnvLsIsEmpty();
```
Did you forget to remove these comments?

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -120,6 +120,11 @@ if (!process.env.CI) {
+
+  if (process.env.VERCEL_SCOPE) {
+    defaultArgs.push('--scope', process.env.VERCEL_SCOPE);
+  }
```
Hmm, I'm not sure this is a good idea since we have tests for `--scope` already.

See the test `try to deploy with non-existing team` for example. We don't want to pass two `--scope` options.

styfle reviewed (commented):
Did you intend to delete `examples/docusaurus-2/yarn.lock`? I think we need that file.

styfle reviewed (commented):
Theres a merge conflict but besides that it looks good 👍

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -536,14 +537,38 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
-    t.is(homeRes.status, 200, formatOutput({ stderr, stdout }));
-    const homeJson = await homeRes.json();
-    t.is(homeJson['MY_ENV_VAR'], 'MY_VALUE');
-    t.is(apiJson['VERCEL_URL'], host);
```
Please add this back. The existing `nowDeployWithVar()` function should not be modified.

MaxLeiter replied:
Whoops, nice catch.

styfle reviewed (dismissed):
Looks great, thanks!

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -546,6 +547,38 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
+    console.log('API URL: ' + localhost[0]);
+    const { host } = new URL(localhost[0]);
+
+    const apiUrl = `http://${host}/api/get-env`;
```
```suggestion
    const apiUrl = `${localhost[0]}/api/get-env`;
```

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -534,22 +535,51 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
+    });
+
+    const apiUrl = `${localhost[0]}/api/get-env`;
+    const apiRes = await fetch(apiUrl);
```
I only see one URL tested here - the API. Lets test the home page which is a static build to ensure the cloud assigns build env vars too.

styfle reviewed (dismissed):
Perfect 👌

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -534,20 +535,55 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
+    const homeRes = await fetch(homeUrl);
+    const homeJson = await homeRes.json();
+    t.is(homeJson['MY_ENV_VAR'], 'MY_VALUE');
+    t.is(homeJson['VERCEL_URL'], localhost[0]);
```
```suggestion
    t.is(homeJson['VERCEL_URL'], '');
```

We don't populate the VERCEL_URL or any system env vars during development

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -534,20 +535,55 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
     t.is(apiJson['VERCEL_URL'], host);
 
     const homeUrl = `https://${host}`;
-    console.log({ homeUrl });
```
Any reason you're removing these logs?

MaxLeiter replied:
Thought they were left in

styfle commented on packages/now-cli/test/integration.js:
```diff
@@ -542,12 +543,50 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
 
     const homeUrl = `https://${host}`;
-    console.log({ homeUrl });
+    console.log({ apiUrl });
```
```suggestion
    console.log({ homeUrl });
```

nkzawa replied:
Looks like this is not applied, why?

lucleray reviewed (commented):
Can we add a test to verify that the behaviour when `now dev` is run and a `.env` file exists is correct?

nkzawa commented on packages/now-cli/src/util/get-decrypted-env-records.ts:
```diff
@@ -0,0 +1,51 @@
+    return {};
+  }
+
+  const envs = await getEnvVariables(output, client, project.id, 4, target);
```
```suggestion
  const envs = await getEnvVariables(output, client, project.id, apiVersion, target);
```
?

nkzawa replied:
Can we make clear how this is solved?

nkzawa replied:
Oh ok, it does not receive the variable on argument anymore.

styfle commented:
@lucleray That test already exists, its just above the one added

nkzawa commented on packages/now-cli/src/commands/env/pull.ts:
```diff
@@ -57,44 +59,29 @@ export default async function pull(
-      })
+  const records: Env = await withSpinner(
+    'Downloading',
+    async () =>
+      await getDecryptedEnvRecords(
+        output,
+        client,
+        project,
+        ProjectEnvTarget.Development
+      )
```
```suggestion
    getDecryptedEnvRecords(
      output,
      client,
      project,
      ProjectEnvTarget.Development
    )
```

nkzawa replied:
Actually, my suggestion is wrong since it takes function instead of a promise.

nkzawa commented on packages/now-cli/src/util/get-decrypted-env-records.ts:
```diff
@@ -0,0 +1,50 @@
+      continue;
+    }
+
+    results[key] = value ? value : '';
```
Just for confirmation, it looks like this does not allow duplicated keys even tho original code used to allow them. Is it no problem?

MaxLeiter replied:
cc @styfle?

styfle replied:
The API prevents multiple keys for the same `target` so this should be fine 

![error-msg](https://user-images.githubusercontent.com/229881/85312212-80ee3900-b484-11ea-94ff-cdd0db451a6e.png)

styfle commented on packages/now-cli/src/commands/env/pull.ts:
```diff
@@ -57,44 +59,29 @@ export default async function pull(
-        return true;
-      })
-      .map(({ key, value }) => `${key}="${escapeValue(value)}"`)
+    Object.keys(records)
+      .map(key => `${key}="${escapeValue(records[key]!)}"`)
```
```suggestion
    Object.entries(records)
      .map([key, value] => `${key}="${escapeValue(value)}"`)
```

nkzawa commented on packages/now-cli/src/commands/dev/dev.ts:
```diff
@@ -93,5 +93,5 @@ export default async function dev(
   process.once('SIGTERM', () => devServer.stop());
 
-  await devServer.start(...listen);
+  await devServer.start(client, link.project, ...listen);
```
I think it's better to pass these arguments to the constructor of  `DevServer` since they're used on `getNowConfig` which is called from other methods as well. (tho it looks working fine because of cache).

nkzawa commented on packages/now-cli/test/integration.js:
```diff
@@ -591,11 +592,86 @@ test('Deploy `api-env` fixture and test `vercel env` command', async t => {
+    t.is(apiJson['MY_ENV_VAR'], 'MY_VALUE');
+    t.is(apiJson['VERCEL_URL'], '');
+
+    const homeUrl = `${localhost[0]}`;
```
nitpick but template literal is not necessary here?

nkzawa reviewed (dismissed):
a few nitpick and suggestion but otherwise LGTM

MaxLeiter commented:
@TooTallNate @nkzawa sorry for dismissed review, had to address a merge conflict. The change since last review was adapting to the new `getNowConfig()` in `server.ts`

MaxLeiter commented on packages/now-cli/test/helpers/prepare.js:
```diff
@@ -365,7 +365,7 @@ CMD ["node", "index.js"]`,
         private: true,
         scripts: {
-          build: 'mkdir public && node print.js > public/index.json',
+          build: 'mkdir -p public && node print.js > public/index.json',
```
this is the change in this file

MaxLeiter committed:
Fetch environment variables from project if no environment file present

MaxLeiter commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -764,7 +803,40 @@ export default class DevServer {
+    const devCommandPromise = this.runDevCommand();
+
+    let address: string | null = null;
+    while (typeof address !== 'string') {
```
this moved up so `address` is available when `VERCEL_URL` is set.

MaxLeiter replied:
This also fixes a bug where if `VERCEL_URL` is set in your `.env` (like from `vc env pull`), it's stored and set to an empty string `VERCEL_URL=''`

<img width="410" alt="Screen Shot 2020-07-20 at 12 51 26 AM" src="https://user-images.githubusercontent.com/8675906/87913055-4ea00f00-ca23-11ea-9e01-f6632ecf4992.png">

MaxLeiter replied:
@TooTallNate

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -511,11 +518,11 @@ export default class DevServer {
   };
 
-  getNowConfig(): Promise<NowConfig> {
+  getNowConfig(client?: Client, project?: Project): Promise<NowConfig> {
```
Why is `client` optional?

Also why is `project` optional? Seems like we already error when no project is linked in `vc dev`.

MaxLeiter replied:
So we don't pass them around each call to getNowConfig

MaxLeiter replied:
They could be moved to ProjectSettings, or something else other than via `start()`?

MaxLeiter committed:
Small cleanup

MaxLeiter committed:
Small cleanup

MaxLeiter committed:
fix circle

MaxLeiter committed:
Move code from out of server to dev

Commit:
Merge branch 'master' into maxleiter/cloudEnvVars

MaxLeiter commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -491,10 +494,7 @@ export default class DevServer {
-        NOW_URL: host,
-        VERCEL_REGION: 'dev1',
-        VERCEL_URL: host,
+        ...this.populateEnvConfig(this.environmentVars, host),
```
I could get the `host` via `this.address` inside populateEnvConfig

styfle commented on packages/now-cli/src/util/get-decrypted-env-records.ts:
```diff
@@ -0,0 +1,50 @@
+): Promise<Env> {
+  if (!project) {
+    return {};
+  }
```
Similarly here?

styfle commented on packages/now-cli/src/util/dev/types.ts:
```diff
@@ -27,11 +27,12 @@ export interface DevServerOptions {
 
 export interface EnvConfigs {
   /**
-   * environment variables from `.env.build` file (deprecated)
```
I'm curious why deprecated was removed?

MaxLeiter replied:
It's no longer deprecated. @TooTallNate and I discussed on Slack a while back

MaxLeiter committed:
Remove useless checks

MaxLeiter committed:
Merge

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -141,11 +141,14 @@ export default class DevServer {
     this.debug = options.debug;
     this.output = options.output;
     this.envConfigs = { buildEnv: {}, runEnv: {}, allEnv: {} };
+    this.environmentVars = options.environmentVars || {};
```
Something seems wrong here. We need the `envConfigs` to `options.environmentVars` if the user didn't provide any files.

MaxLeiter replied:
Happens in getNowConfig

MaxLeiter commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -764,7 +783,39 @@ export default class DevServer {
+      .replace('127.0.0.1', 'localhost');
+
     const nowConfig = await this.getNowConfig();
+    const devCommandPromise = this.runDevCommand();
```
@TooTallNate moved the init to below starting the server, and left the await at the end

MaxLeiter committed:
Better function name

MaxLeiter committed:
Populate env vars only if provided in .env, and fetch/populate vars from
cloud if no .env

MaxLeiter committed:
remove log

MaxLeiter committed:
Only need host with VERCEL_URL

MaxLeiter committed:
Add MY_ENV_VAR back to test

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -630,13 +626,20 @@ export default class DevServer {
+    let allEnv = { ...buildEnv, ...runEnv };
+
+    // If no .env/.build.env is present, fetch and use cloud environment variables
+    if (Object.keys(allEnv).length === 0) {
```
I think this will never be true because `NOW_REGION` is always added, right?

styfle replied:
I would suggest adding `NOW_REGION` to `runEnv` after reading the files and reading the cloudEnv.

Something like this:

```js
if (Object.keys(allEnv).length === 0) {
  const cloudEnv = this.populateVercelEnvVars(this.environmentVars);
  runEnv = buildEnv = cloudEnv;
}
runEnv['NOW_REGION'] = 'dev1';
const allEnv = { ...buildEnv, ...runEnv };
this.envConfigs = { buildEnv, runEnv, allEnv };
```

styfle replied:
Nevermind, ignore my comments

MaxLeiter committed:
envVars shouldn't be {}

MaxLeiter committed:
add startPromise

MaxLeiter committed:
Only allow start() once

TooTallNate committed:
Return `startPromise`

kodiakhq[bot] committed:
Merge branch 'master' into maxleiter/cloudEnvVars

styfle commented on packages/now-cli/src/util/dev/server.ts:
```diff
@@ -741,6 +746,26 @@ export default class DevServer {
+        const host = new URL(this.address).host;
+        env['VERCEL_URL'] = host;
+      } else if (name === 'VERCEL_REGION') {
+        env['VERCEL_REGION'] = 'dev1';
```
@TooTallNate Do we want to use `dev` instead of `dev1` like leo suggested in [ch4611](https://app.clubhouse.io/vercel/story/4611)? Once we ship this, I don't think we can change it.

TooTallNate replied:
I mean I explained my opinion about it in the ch story, but ultimately I don't think it's my call. If we do change it to `dev` we'll need to update [the documentation](https://vercel.com/docs/v2/edge-network/regions#routing) as well.

TooTallNate committed:
Use "vercel-bash@3.0.8"

MaxLeiter committed:
add .env to node-ts-node-target

MaxLeiter committed:
Merge branch 'maxleiter/cloudEnvVars' of github.com:vercel/vercel into maxleiter/cloudEnvVars

MaxLeiter committed:
Remove NOW_URL check

kodiakhq[bot] committed:
Merge branch 'master' into maxleiter/cloudEnvVars
